src/libsync/filesystem.cpp

```diff
@@ -198,5 +207,100 @@ bool FileSystem::getInode(const QString &filename, quint64 *inode)
     return false;
 }
 
+namespace {
+
+#ifdef Q_OS_LINUX
+    Q_ALWAYS_INLINE ssize_t getxattr(const char *path, const char *name, void *value, size_t size, u_int32_t, int)
+    {
+        return ::getxattr(path, name, value, size);
+    }
+
+    Q_ALWAYS_INLINE int setxattr(const char *path, const char *name, const void *value, size_t size, u_int32_t, int)
+    {
+        return ::setxattr(path, name, value, size, 0);
+    }
+
+    Q_ALWAYS_INLINE int removexattr(const char *path, const char *name, int)
+    {
+        return ::removexattr(path, name);
+    }
+#endif // Q_OS_LINUX
+
+} // anonymous namespace
+
+std::optional<QByteArray> FileSystem::Tags::get(const QString &path, const QString &key)
+{
+#if defined(Q_OS_MAC) || defined(Q_OS_LINUX)
+    QString platformKey = key;
+    if (Utility::isLinux()) {
+        platformKey = QStringLiteral("user.") + platformKey;
+    }
+
+    QByteArray value(MaxValueSize + 1, '\0'); // Add a NUL character to terminate a string
+    auto size = getxattr(path.toUtf8().constData(), platformKey.toUtf8().constData(), value.data(), MaxValueSize, 0, 0);
+    if (size != -1) {
+        value.truncate(size);
+        return value;
+    }
+#elif defined(Q_OS_WIN)
+    QFile file(QStringLiteral("%1:%2").arg(path, key));
+    if (file.open(QIODevice::ReadOnly)) {
+        return file.readAll();
+    }
+#endif // Q_OS_MAC || Q_OS_LINUX
+
+    return {};
+}
+
+OCC::Result<void, QString> FileSystem::Tags::set(const QString &path, const QString &key, const QByteArray &value)
+{
+    OC_ASSERT(value.size() < MaxValueSize)
```

Comment:
Should we error on implicit overrides? Or would that only make sense for setDirectoryTag?

Reply:
Neither? This is when one sync app is removed and another one is installed. Then the new application can take over the old sync root. Or is this a bad idea?

Reply:
The assert would only occur in dev setups.
For the above assert a Q_ASSERT should be sufficient.

I'd say a normal scenario would be that a sync root is properly removed, so we also removed the tag?
Maybe not an assert, but a general warning as the folder was not properly cleaned?

Reply:
Done